resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java

```diff
@@ -197,79 +212,64 @@ private void queue(AsyncOperation op)
          // fetch it from the context directly to avoid having to restore the context just in case we're invoked on a context-less thread
          HttpRequest resteasyRequest = (HttpRequest) contextDataMap.get(HttpRequest.class);
          if(request.isAsyncStarted() && !resteasyRequest.getAsyncContext().isOnInitialRequest()) {
+            boolean flush = false;
             synchronized(this) {
-               ServletOutputStream os;
-               try
-               {
-                  os = response.getOutputStream();
-               } catch (IOException e)
-               {
-                  // return a failed future, do not queue it
-                  op.future.completeExceptionally(e);
-                  return;
+               if (asyncRegistered.compareAndSet(false, true)) {
+                  out.setWriteListener(this);
                }
-               if(!asyncRegistered) {
-                  // start the queue
-                  asyncRegistered = true;
-                  // make sure we have something ready to be executed
-                  addToQueue(op);
-                  os.setWriteListener(this);
-                  // never call isReady before Undertow is ready and has already called our listener at least once
-               } else if(asyncListenerCalled && os.isReady()) {
+               if(asyncListenerCalled && out.isReady()) {
                   // it's possible that we startAsync and queue, then queue another event and the stream becomes ready before
                   // onWritePossible is called, which means we need to flush the queue here to guarantee ordering if that happens
-                  addToQueue(op);
-                  flushQueue(os);
+                  asyncQueue.add(op);
+                  flush = true;
                } else {
                   // just queue
-                  addToQueue(op);
+                  asyncQueue.add(op);
                }
             }
+            // Invoked outside the lock to avoid deadlocks, the flushQueue itself locks on this
+            if (flush) {
+               flushQueue();
+            }
```

Comment:
The `flushQueue()` itself blocks on `this` which is why it's outside of the lock. This is just so adding to the queue will release quicker. In the `SseEventOutputImpl` I went with one of your original approaches of using the `AsyncOutputStream` from the `HttpResponse` as the lock. It feels a bit strange, but it's the right thing to do to avoid the deadlock.

We can definitely look at this again though. It would be nice to not lock as much, but we just need to ensure the order is preserved.